tests/integration/abi_router_test.py

```diff
@@ -0,0 +1,464 @@
+    ),
+    (
+        qa_runner("mod"),
+        [TYPICAL_IAC_OC],
+        qc_runner("mod"),
+        [],
+        {DRProp.passed: True, DRProp.lastLog: lambda args: args[1] % args[2]},
+    ),
+    (
+        qa_runner("all_laid_to_args"),
+        [TYPICAL_IAC_OC],
+        qc_runner("all_laid_to_args"),
+        [],
+        {DRProp.passed: True, DRProp.lastLog: lambda args: sum(args[1:])},
+    ),
+    (
+        qa_runner("empty_return_subroutine"),
+        [
+            (False, OnComplete.NoOpOC),
+            (False, OnComplete.OptInOC),
+            (True, OnComplete.OptInOC),
+        ],
+        qc_runner("empty_return_subroutine"),
+        [
+            (False, OnComplete.ClearStateOC),
+        ],
+        {
+            DRProp.passed: True,
+            DRProp.lastLog: DryRunEncoder.hex(
+                "appear in both approval and clear state"
+            ),
+        },
+    ),
+    (
+        qa_runner("log_1"),
+        [(False, OnComplete.NoOpOC), (False, OnComplete.OptInOC)],
+        qc_runner("log_1"),
+        [
+            (False, OnComplete.ClearStateOC),
+        ],
+        {DRProp.passed: True, DRProp.lastLog: 1},
+    ),
+    (
+        qa_runner("log_creation"),
+        [(True, OnComplete.NoOpOC)],
+        qc_runner("log_creation"),
+        [],
+        {DRProp.passed: True, DRProp.lastLog: "logging creation"},
+    ),
+    (
+        qa_runner(
+            "approve_if_odd"
+        ),  # this should only appear in the clear-state program
+        [],
+        qc_runner("approve_if_odd"),
+        [
+            (False, OnComplete.ClearStateOC),
+        ],
+        {
+            DRProp.passed: lambda args: args[1] % 2 == 1,
+            DRProp.lastLog: None,
+        },
+    ),
+    (
+        qa_runner(None),
+        [(False, OnComplete.OptInOC)],
+        qc_runner(None),
+        [
+            (False, OnComplete.ClearStateOC),
+        ],
+        {
+            DRProp.passed: True,
+            DRProp.lastLog: lambda _, actual: actual
+            in (None, DryRunEncoder.hex("optin call")),
+        },
+    ),
+]
+
+
+def make_yacc_runner(method, approval):
+    return MethodRunner(
+        f"yacc_{'approval' if approval else 'clear'}",
+        approval,
+        YACC_TEAL if approval else YACC_CLEAR_TEAL,
+        CONTRACT,
+        method,
+    )
+
+
+# strip out the bare calls for YACC ~ "yetAnotherContractConstructedFromRouter":
+def yaccify(q_cases):
+    return [
+        (make_yacc_runner(method, True), c[1], make_yacc_runner(method, False), *c[3:])
+        for c in q_cases
+        if (method := c[0].method)
+    ]
+
+
+YACC_CASES = yaccify(QUESTIONABLE_CASES)
+
+ALL_CASES = QUESTIONABLE_CASES + YACC_CASES
+
+
+def get_aa_strat(method_runner, abi_args_mod=None) -> ABICallStrategy:
+    return ABICallStrategy(
+        method_runner.contract,
+        RandomABIStrategyHalfSized,
+        num_dryruns=NUM_ROUTER_DRYRUNS,
+        abi_args_mod=abi_args_mod,
+    )
```

Comment:
## STEP 2. Provide an `ABICallStrategy` generating dry runs inputs for each call
It needs:
* an ABI JSON `contract`
* an `ABIStrategy` _type_ to generate each individual arg
* `num_dryruns` for the number of input runs to generate
* `abi_args_mod` - possibly an `ABIArgsMod` **enum** that specifies how to _mutate_ the tweak for exploring what happens outside of the boundaries of anticipated inputs